tonic/src/transport/server/mod.rs

```diff
@@ -25,26 +25,33 @@ use crate::transport::Error;
 use self::recover_error::RecoverError;
 use super::service::{GrpcTimeout, Or, Routes, ServerIo};
 use crate::{body::BoxBody, request::ConnectionInfo};
+use bytes::Bytes;
 use futures_core::Stream;
 use futures_util::{
-    future::{self, Either as FutureEither, MapErr},
-    TryFutureExt,
+    future::{self, MapErr},
+    ready, TryFutureExt,
 };
 use http::{Request, Response};
+use http_body::Body as _;
 use hyper::{server::accept, Body};
+use pin_project::pin_project;
 use std::{
     fmt,
     future::Future,
     net::SocketAddr,
+    pin::Pin,
     sync::Arc,
     task::{Context, Poll},
     time::Duration,
 };
 use tokio::io::{AsyncRead, AsyncWrite};
-use tower::{limit::concurrency::ConcurrencyLimitLayer, util::Either, Service, ServiceBuilder};
-use tracing_futures::{Instrument, Instrumented};
+use tower::{
+    layer::util::Identity, layer::Layer, limit::concurrency::ConcurrencyLimitLayer, util::Either,
+    Service, ServiceBuilder,
+};
 
-type BoxService = tower::util::BoxService<Request<Body>, Response<BoxBody>, crate::Error>;
+type BoxHttpBody = http_body::combinators::BoxBody<Bytes, crate::Error>;
```

Comment:
`tonic::body::BoxBody` uses `Status` as the error type which break tower middleware that wraps the response body, as is common in tower-http. Have to use this different error type for response bodies instead since it supports any error. Its only. used in the `serve_*` methods and not in `add_service` and `add_optional_service`.

Reply:
Requiring the `Data` type to be `Bytes` should be fine and simplifies parts of the implementation.

Reply:
Yeah I think this sounds right